In [5]:
#importing lybriers
import pandas as pd
import numpy as np

In [6]:
#creating path
path=r"C:\Users\SAGAR\Downloads\insurance.csv"

In [7]:
#reading of file using path
mdf=pd.read_csv(path)

In [8]:
#top five rows 
mdf.head()

age     sex     bmi  children smoker     region      charges
0   19  female  27.900         0    yes  southwest  16884.92400
1   18    male  33.770         1     no  southeast   1725.55230
2   28    male  33.000         3     no  southeast   4449.46200
3   33    male  22.705         0     no  northwest  21984.47061
4   32    male  28.880         0     no  northwest   3866.85520

In [9]:
#spliting of data into x & y (input & output)
y=mdf['charges']
X=mdf[['age','sex','bmi','children','smoker','region']]

In [10]:
#top 5 rows of input 
X.head()

age     sex     bmi  children smoker     region
0   19  female  27.900         0    yes  southwest
1   18    male  33.770         1     no  southeast
2   28    male  33.000         3     no  southeast
3   33    male  22.705         0     no  northwest
4   32    male  28.880         0     no  northwest

In [11]:
#top 5 rows of output
y.head()

0    16884.92400
1     1725.55230
2     4449.46200
3    21984.47061
4     3866.85520
Name: charges, dtype: float64

In [12]:
from sklearn.model_selection import train_test_split # importing of model selection
X_train, X_test, y_train, y_test = train_test_split(X,y,train_size = 0.75,random_state=0)  # spliting of data in to four parts

print(X_train.shape,y_train.shape)
print(X_test.shape,y_test.shape)

(1003, 6) (1003,)
(335, 6) (335,)


In [13]:
X.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   object 
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   object 
 5   region    1338 non-null   object 
dtypes: float64(1), int64(2), object(3)
memory usage: 62.8+ KB


In [14]:
#saperating of numerical data from X_train
X_train_num = X_train.select_dtypes(include=['int64','float64']) 
X_train_num.head()

age    bmi  children
1075   32  29.59         1
131    61  22.04         0
15     19  24.60         1
1223   20  24.42         0
1137   26  22.23         0

In [15]:
#saperating of catagorical data from X_train
X_train_cat = X_train.select_dtypes(include=['object'])
X_train_cat.head()

sex smoker     region
1075  female     no  southeast
131   female     no  northeast
15      male     no  southwest
1223  female    yes  southeast
1137  female     no  northwest

In [16]:
from sklearn.preprocessing import StandardScaler  # importing of standardscaler from sklearn.preprocessing
scaler = StandardScaler()
X_train_num_transformed = pd.DataFrame(scaler.fit_transform(X_train_num), # transforming of numerical data and creating of dataframe for transformed data
                                       columns= X_train_num.columns,
                                       index=X_train_num.index)
X_train_num_transformed.head()

age       bmi  children
1075 -0.514853 -0.181331 -0.063607
131   1.548746 -1.393130 -0.892144
15   -1.439915 -0.982242 -0.063607
1223 -1.368757 -1.011133 -0.892144
1137 -0.941805 -1.362635 -0.892144

In [17]:
from sklearn.preprocessing import OneHotEncoder # importing of OneHotEncoder
encoder = OneHotEncoder(drop='first',sparse=False)
X_train_cat_ohe = pd.DataFrame(encoder.fit_transform(X_train_cat),
                               columns = encoder.get_feature_names_out(X_train_cat.columns),
                               index = X_train_cat.index) #transforming of catagorical data and creating of dataframe for transformed data
X_train_cat_ohe.head()

sex_male  smoker_yes  region_northwest  region_southeast  \
1075       0.0         0.0               0.0               1.0   
131        0.0         0.0               0.0               0.0   
15         1.0         0.0               0.0               0.0   
1223       0.0         1.0               0.0               1.0   
1137       0.0         0.0               1.0               0.0   

      region_southwest  
1075               0.0  
131                0.0  
15                 1.0  
1223               0.0  
1137               0.0

In [77]:
#concataning the neumerical transformed data and catogorical transformed data
X_train_transformed= pd.concat([X_train_num_transformed,X_train_cat_ohe],axis=1) 
X_test_transformed.head()

age       bmi  children  sex_male  smoker_yes  region_northwest  \
578   0.908319 -0.083424 -0.063607       1.0         0.0               0.0   
610   0.552526 -0.216642 -0.063607       0.0         0.0               0.0   
569   0.623684  1.580192  0.764931       1.0         1.0               1.0   
1034  1.548746  1.229492 -0.892144       1.0         0.0               1.0   
198   0.837160 -2.033538 -0.892144       0.0         0.0               1.0   

      region_southeast  region_southwest  
578                0.0               1.0  
610                1.0               0.0  
569                0.0               0.0  
1034               0.0               0.0  
198                0.0               0.0

In [19]:
X_test.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 335 entries, 578 to 1337
Data columns (total 6 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       335 non-null    int64  
 1   sex       335 non-null    object 
 2   bmi       335 non-null    float64
 3   children  335 non-null    int64  
 4   smoker    335 non-null    object 
 5   region    335 non-null    object 
dtypes: float64(1), int64(2), object(3)
memory usage: 18.3+ KB


In [20]:
# saperating of numerical data from X_test
X_test_num = X_test.select_dtypes(include=['int64','float64'])
X_test_num.head()

age     bmi  children
578    52  30.200         1
610    47  29.370         1
569    48  40.565         2
1034   61  38.380         0
198    51  18.050         0

In [21]:
#saparating of catagorical data from X_test
X_test_cat = X_test.select_dtypes(include=['object'])
X_test_cat.head()

sex smoker     region
578     male     no  southwest
610   female     no  southeast
569     male    yes  northwest
1034    male     no  northwest
198   female     no  northwest

In [22]:
# transforming of numerical data and creating of dataframe for transformed data
X_test_num_rescaled= pd.DataFrame(scaler.transform(X_test_num),
                                 columns= X_test_num.columns,
                                 index= X_test_num.index)
X_test_num_rescaled.head()

age       bmi  children
578   0.908319 -0.083424 -0.063607
610   0.552526 -0.216642 -0.063607
569   0.623684  1.580192  0.764931
1034  1.548746  1.229492 -0.892144
198   0.837160 -2.033538 -0.892144

In [23]:
#transforming of catagorical data and creating of dataframe for transformed data
X_test_cat_ohe = pd.DataFrame(encoder.transform(X_test_cat),
                               columns = encoder.get_feature_names_out(X_test_cat.columns),
                               index = X_test_cat.index)
X_test_cat_ohe.head()

sex_male  smoker_yes  region_northwest  region_southeast  \
578        1.0         0.0               0.0               0.0   
610        0.0         0.0               0.0               1.0   
569        1.0         1.0               1.0               0.0   
1034       1.0         0.0               1.0               0.0   
198        0.0         0.0               1.0               0.0   

      region_southwest  
578                1.0  
610                0.0  
569                0.0  
1034               0.0  
198                0.0

In [24]:
#concataning the neumerical transformed data and catogorical transformed data
X_test_transformed= pd.concat([X_test_num_rescaled,X_test_cat_ohe],axis=1)
X_test_transformed.head()

age       bmi  children  sex_male  smoker_yes  region_northwest  \
578   0.908319 -0.083424 -0.063607       1.0         0.0               0.0   
610   0.552526 -0.216642 -0.063607       0.0         0.0               0.0   
569   0.623684  1.580192  0.764931       1.0         1.0               1.0   
1034  1.548746  1.229492 -0.892144       1.0         0.0               1.0   
198   0.837160 -2.033538 -0.892144       0.0         0.0               1.0   

      region_southeast  region_southwest  
578                0.0               1.0  
610                1.0               0.0  
569                0.0               0.0  
1034               0.0               0.0  
198                0.0               0.0

In [48]:
# preparation of model using algorithum
from sklearn.linear_model import LinearRegression
LR_regressor = LinearRegression()
LR_regressor.fit(X_train_transformed,y_train)

LinearRegression()

In [57]:
# prediction of values
y_test_prdt= LR_regressor.predict(X_test_transformed)

In [58]:
# creating of dataframe for actual values and predicted values 
tem_df= pd.DataFrame({'Actual':y_test,'predict':y_test_prdt})
tem_df.head()

Actual       predict
578    9724.53000  11121.101409
610    8547.69130   9369.083452
569   45702.02235  38349.258807
1034  12950.07120  16331.935006
198    9644.25250   7041.227706

In [59]:
# calculating of Mean absoult error, Mean squared error and Root Mean absoult error
from sklearn import metrics
print('Mean absoult error', metrics.mean_absolute_error(y_test,y_test_prdt))
print('Mean squared error', metrics.mean_squared_error(y_test,y_test_prdt))
print('Root Mean absoult error', np.sqrt(metrics.mean_absolute_error(y_test,y_test_prdt)))

Mean absoult error 4012.7152768530864
Mean squared error 32135747.23938177
Root Mean absoult error 63.345996533743836


In [60]:
## preparation of model using algorithum
from sklearn.neighbors import KNeighborsRegressor
KN_regressor = KNeighborsRegressor()
KN_regressor.fit(X_train_transformed, y_train)

KNeighborsRegressor()

In [61]:
# prediction of values
y_test_prdt= KN_regressor.predict(X_test_transformed)

In [62]:
# creating of dataframe for actual values and predicted values 
tem1_df= pd.DataFrame({'Actual':y_test,'predict':y_test_prdt})
tem1_df.head()

Actual       predict
578    9724.53000   8433.604400
610    8547.69130   7948.160220
569   45702.02235  29049.586058
1034  12950.07120  12094.782330
198    9644.25250  11108.725660

In [63]:
# calculating of Mean absoult error, Mean squared error and Root Mean absoult error
from sklearn import metrics
print('Mean absoult error', metrics.mean_absolute_error(y_test,y_test_prdt))
print('Mean squared error', metrics.mean_squared_error(y_test,y_test_prdt))
print('Root Mean absoult error', np.sqrt(metrics.mean_absolute_error(y_test,y_test_prdt)))

Mean absoult error 4146.4536904686565
Mean squared error 45611979.05104784
Root Mean absoult error 64.39296305085406


In [64]:
## preparation of model using algorithum
from sklearn.tree import DecisionTreeRegressor
DT_regressor = DecisionTreeRegressor()
DT_regressor.fit(X_train_transformed , y_train)

DecisionTreeRegressor()

In [65]:
# prediction of values
y_test_prdt= DT_regressor.predict(X_test_transformed)

In [66]:
# creating of dataframe for actual values and predicted values 
tem2_df= pd.DataFrame({'Actual':y_test,'predict':y_test_prdt})
tem2_df.head()

Actual      predict
578    9724.53000  10085.84600
610    8547.69130   8233.09750
569   45702.02235  44202.65360
1034  12950.07120  12741.16745
198    9644.25250   9855.13140

In [67]:
# calculating of Mean absoult error, Mean squared error and Root Mean absoult error
from sklearn import metrics
print('Mean absoult error', metrics.mean_absolute_error(y_test,y_test_prdt))
print('Mean squared error', metrics.mean_squared_error(y_test,y_test_prdt))
print('Root Mean absoult error', np.sqrt(metrics.mean_absolute_error(y_test,y_test_prdt)))

Mean absoult error 3385.3565601761197
Mean squared error 48990553.60614237
Root Mean absoult error 58.18381699558839


In [68]:
## preparation of model using algorithum
from sklearn.ensemble import RandomForestRegressor
RF_regressor = RandomForestRegressor()
RF_regressor.fit(X_train_transformed,y_train)

RandomForestRegressor()

In [69]:
# prediction of values
y_test_prdt= RF_regressor.predict(X_test_transformed)

In [70]:
# creating of dataframe for actual values and predicted values 
tem3_df= pd.DataFrame({'Actual':y_test,'predict':y_test_prdt})
tem3_df.head()

Actual       predict
578    9724.53000  10388.708366
610    8547.69130   9814.805761
569   45702.02235  45760.830256
1034  12950.07120  13332.136206
198    9644.25250   9773.063055

In [71]:
# calculating of Mean absoult error, Mean squared error and Root Mean absoult error
from sklearn import metrics
print('Mean absoult error', metrics.mean_absolute_error(y_test,y_test_prdt))
print('Mean squared error', metrics.mean_squared_error(y_test,y_test_prdt))
print('Root Mean absoult error', np.sqrt(metrics.mean_absolute_error(y_test,y_test_prdt)))

Mean absoult error 2699.198941297264
Mean squared error 19896814.54858533
Root Mean absoult error 51.95381546428774


In [72]:
## preparation of model using algorithum
from sklearn.ensemble import GradientBoostingRegressor
GBDT_regressor = GradientBoostingRegressor()
GBDT_regressor.fit(X_train_transformed, y_train)

GradientBoostingRegressor()

In [73]:
# prediction of values
y_test_prdt= GBDT_regressor.predict(X_test_transformed)

In [74]:
# creating of dataframe for actual values and predicted values 
tem4_df= pd.DataFrame({'Actual':y_test,'predict':y_test_prdt})
tem4_df.head()

Actual       predict
578    9724.53000  12000.468696
610    8547.69130   9496.052685
569   45702.02235  46238.972154
1034  12950.07120  14502.929530
198    9644.25250  10693.365565

In [75]:
# calculating of Mean absoult error, Mean squared error and Root Mean absoult error
print('Mean absoult error', metrics.mean_absolute_error(y_test,y_test_prdt))
print('Mean squared error', metrics.mean_squared_error(y_test,y_test_prdt))
print('Root Mean absoult error', np.sqrt(metrics.mean_absolute_error(y_test,y_test_prdt)))

Mean absoult error 2427.7320391641847
Mean squared error 16105653.713913523
Root Mean absoult error 49.27202085529053


# Observation

* Linear Regression Mean absoult error____________: 4012.7152768530864
* KNN regresssion Mean absoult error _____________: 4146.4536904686565
* Decision Tree regression Mean absoult error____: 3385.3565601761197
* Random Forest Regression Mean absoult error____: 2699.198941297264
* GradientBoostingRegressor Mean absoult error____: 2427.7320391641847

# Conclusion

* From the above observations the GradientBoostingRegressor algorithum is the best for medical cost prediction because it has
  least mean absoult error